src/fsharp/FSharp.Core/control.fs

```diff
@@ -896,7 +915,11 @@ namespace Microsoft.FSharp.Control
             queueAsync 
                     token                        
                     (fun res -> resultCell.RegisterResult(Ok(res),reuseThread=true))
-                    (fun edi -> resultCell.RegisterResult(Error(edi),reuseThread=true))
+                    (fun edi ->
+                        let result =
+                            if token.IsCancellationRequested then Canceled(protectExn (new OperationCanceledException()) edi)
```

Comment:
Why do we need this ``token.IsCancellationRequested`` check so late.  There is always a race here - we've reached the end of the computation and found an Error result. Why another check for cancellation so late?

I mean, it's certainly valid to check for cancellation so late, but is it necessary for the PR?

Reply:
I think this is to have a consistent result. If I remember currently we can have different results depending of whether our `Async` contained a task or not. Or even worse how the code is formatted.
For example adding a `try - with` could lead to a different exception (even if the `with` block has nothing to do with the actual exception thrown). I think one of the tests actually checks this.

So basically this is to give a more consistent result when there is an interop with tasks

Reply:
I don't yet see how this achieves consistency.  `IsCancellationRequested` could become true the moment after we check it. And I don't think there are cases where we know a-priori that it was true coming into the exception continuation (we should never be taking the exception continuation once the cancellation continuation is taken)

I do understand that there is a case related to tasks where you want to make things consistent

> For example adding a try - with could lead to a different exception (even if the with block has nothing to do with the actual exception thrown). 

Could you give me an example?

Reply:
So the following test will fail:

```fsharp
    [<Test>]
    member this.StartAsTaskCancellationViaException () =
        let cts = new CancellationTokenSource()
        let tcs = TaskCompletionSource<unit>()
        let a = async {
            cts.CancelAfter (100)
            do! tcs.Task |> Async.AwaitTask }
        use t : Task<unit> =
            Async.StartAsTask(a, cancellationToken = cts.Token)

        // Should not finish
        try
            let result = t.Wait(300)
            Assert.IsFalse (result)
        with :? AggregateException -> Assert.Fail "Task should not finish, yet"

        let msg = "Custom non-conforming 3rd-Party-Api throws"
        tcs.SetException(Exception msg)

        try
            this.WaitASec t
        with :? TaskCanceledException as t ->
            match t.InnerException with
            | :? AggregateException as a ->
                Assert.AreEqual(1, a.InnerExceptions.Count)
                Assert.AreEqual(msg, a.InnerException.Message)
            | _ -> reraise()
        Assert.IsTrue (t.IsCompleted, "Task is not completed")
        Assert.IsTrue (t.IsCanceled, "Task is not cancelled")
```

Basically without checking the cancellation-token check above we will yield no `TaskCanceledException` but an `Exception`. This is inconsistent because if you change the `async` to

```fsharp
        let a = async {
            try
                cts.CancelAfter (100)
                do! tcs.Task |> Async.AwaitTask
            with _ -> printfn "test" }
```

You will in-fact get the `TaskCanceledException` as you would expect. This is surprising behavior at the very least (ie we get a different exception only by adding a `try ... with _ when false` while everything else stays the same).

Therefore I added the additional checks on all the constructs. Is there a better option?